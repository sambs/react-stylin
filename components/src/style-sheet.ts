import { createStyleSheet } from '@sambs/react-stylin'
import { StyleContext } from './context'

export const StyleSheet = createStyleSheet(StyleContext)
