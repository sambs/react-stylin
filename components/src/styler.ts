import { createStyler } from 'stylin'
import { StyleContext } from './context'

export const Styler = createStyler(StyleContext)
