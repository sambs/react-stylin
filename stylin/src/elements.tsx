import * as React from 'react'
import { StyleParams, StyleResolver, useStylin } from './core'

type Elements = {
  a: HTMLAnchorElement
  abbr: HTMLElement
  address: HTMLElement
  area: HTMLAreaElement
  article: HTMLElement
  aside: HTMLElement
  audio: HTMLAudioElement
  b: HTMLElement
  base: HTMLBaseElement
  bdi: HTMLElement
  bdo: HTMLElement
  big: HTMLElement
  blockquote: HTMLElement
  body: HTMLBodyElement
  br: HTMLBRElement
  button: HTMLButtonElement
  canvas: HTMLCanvasElement
  caption: HTMLElement
  cite: HTMLElement
  code: HTMLElement
  col: HTMLTableColElement
  colgroup: HTMLTableColElement
  data: HTMLDataElement
  datalist: HTMLDataListElement
  dd: HTMLElement
  del: HTMLElement
  details: HTMLElement
  dfn: HTMLElement
  dialog: HTMLDialogElement
  div: HTMLDivElement
  dl: HTMLDListElement
  dt: HTMLElement
  em: HTMLElement
  embed: HTMLEmbedElement
  fieldset: HTMLFieldSetElement
  figcaption: HTMLElement
  figure: HTMLElement
  footer: HTMLElement
  form: HTMLFormElement
  h1: HTMLHeadingElement
  h2: HTMLHeadingElement
  h3: HTMLHeadingElement
  h4: HTMLHeadingElement
  h5: HTMLHeadingElement
  h6: HTMLHeadingElement
  head: HTMLHeadElement
  header: HTMLElement
  hgroup: HTMLElement
  hr: HTMLHRElement
  html: HTMLHtmlElement
  i: HTMLElement
  iframe: HTMLIFrameElement
  img: HTMLImageElement
  input: HTMLInputElement
  ins: HTMLModElement
  kbd: HTMLElement
  keygen: HTMLElement
  label: HTMLLabelElement
  legend: HTMLLegendElement
  li: HTMLLIElement
  link: HTMLLinkElement
  main: HTMLElement
  map: HTMLMapElement
  mark: HTMLElement
  menu: HTMLElement
  menuitem: HTMLElement
  meta: HTMLMetaElement
  meter: HTMLElement
  nav: HTMLElement
  noindex: HTMLElement
  noscript: HTMLElement
  object: HTMLObjectElement
  ol: HTMLOListElement
  optgroup: HTMLOptGroupElement
  option: HTMLOptionElement
  output: HTMLElement
  p: HTMLParagraphElement
  param: HTMLParamElement
  picture: HTMLElement
  pre: HTMLPreElement
  progress: HTMLProgressElement
  q: HTMLQuoteElement
  rp: HTMLElement
  rt: HTMLElement
  ruby: HTMLElement
  s: HTMLElement
  samp: HTMLElement
  slot: HTMLSlotElement
  script: HTMLScriptElement
  section: HTMLElement
  select: HTMLSelectElement
  small: HTMLElement
  source: HTMLSourceElement
  span: HTMLSpanElement
  strong: HTMLElement
  style: HTMLStyleElement
  sub: HTMLElement
  summary: HTMLElement
  sup: HTMLElement
  table: HTMLTableElement
  template: HTMLTemplateElement
  tbody: HTMLTableSectionElement
  td: HTMLTableDataCellElement
  textarea: HTMLTextAreaElement
  tfoot: HTMLTableSectionElement
  th: HTMLTableHeaderCellElement
  thead: HTMLTableSectionElement
  time: HTMLElement
  title: HTMLTitleElement
  tr: HTMLTableRowElement
  track: HTMLTrackElement
  u: HTMLElement
  ul: HTMLUListElement
  var: HTMLElement
  video: HTMLVideoElement
  wbr: HTMLElement
  webview: HTMLWebViewElement
}

type ElementName = keyof Elements

type ElementProps<E extends ElementName> = Omit<
  JSX.IntrinsicElements[E],
  'ref' | 'style'
>

type StylinComponentProps<
  E extends ElementName,
  C,
  P extends StyleParams
> = ElementProps<E> & {
  styles?: StyleResolver<C, P>
  innerRef?: React.Ref<Elements[E]>
} & Partial<P>

type CreateStylinComponentParams<
  E extends ElementName,
  C,
  P extends StyleParams
> = {
  defaultProps?: ElementProps<E>
  defaultStyles?: StyleResolver<C, P>
  defaultStyleProps?: P
  displayName?: string
  element: E
  context: React.Context<C>
}

export const createStylinComponent = <
  E extends ElementName,
  C,
  P extends StyleParams
>({
  defaultProps,
  defaultStyles,
  defaultStyleProps,
  displayName = 'StylinComponent',
  element,
  context,
}: CreateStylinComponentParams<E, C, P>) => {
  const component: React.FC<StylinComponentProps<E, C, P>> = (props) => {
    const styleProps: Partial<P> = {}
    const elemProps: Partial<React.ComponentProps<E>> = {}

    for (const prop in defaultStyleProps) {
      if (prop in props) {
        styleProps[prop] = props[prop]
      } else {
        styleProps[prop] = defaultStyleProps[prop]
      }
    }

    for (const prop in props) {
      if (defaultStyleProps && defaultStyleProps.hasOwnProperty(prop)) continue
      if (prop == 'styles' || prop == 'innerRef') continue
      // @ts-ignore
      elemProps[prop] = props[prop]
    }

    const style = useStylin(
      context,
      styleProps as Required<P>,
      defaultStyles,
      props.styles
    )

    return React.createElement(element, {
      ...defaultProps,
      ...elemProps,
      ref: props.innerRef,
      style,
    })
  }

  component.displayName = displayName

  return component
}
