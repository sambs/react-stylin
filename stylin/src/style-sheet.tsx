import * as React from 'react'

export type StyleSheetResolver<C> = (context: C) => string

export type StyleSheetProps<C> = {
  styles: StyleSheetResolver<C>
}

export const createStyleSheet = <C,>(context: React.Context<C>) => {
  const StyleSheet = ({ styles }: StyleSheetProps<C>) => {
    const contextValue = React.useContext(context)
    const rules = styles(contextValue)
    return <style dangerouslySetInnerHTML={{ __html: rules }} />
  }
  return StyleSheet
}
