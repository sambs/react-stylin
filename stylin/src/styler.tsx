import * as React from 'react'
import { StyleContextType, StyleResolver, useStylin } from './core'

export type StylerProps<C extends StyleContextType = StyleContextType> = {
  children: (style: React.CSSProperties) => JSX.Element
  styles: StyleResolver<{}, C>
}

export const Styler = ({ children, styles }: StylerProps) => {
  const style = useStylin({}, styles)
  return children(style)
}
