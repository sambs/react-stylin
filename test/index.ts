import './elements'
import './style-sheet'
import './styler'
