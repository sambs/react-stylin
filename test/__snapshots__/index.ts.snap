// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element with default props 1`] = `
<div
  role="alert"
  style={
    Object {
      "background": "red",
    }
  }
>
  Uh Oh!
</div>
`;

exports[`Element with ref forwarding 1`] = `
<div
  style={
    Object {
      "fontSize": "16px",
    }
  }
>
  3
</div>
`;

exports[`Element with style props 1`] = `
<div
  style={
    Object {
      "fontSize": "16px",
    }
  }
>
  Hey there!
</div>
`;

exports[`Element with style props 2`] = `
<div
  style={
    Object {
      "fontSize": "12px",
    }
  }
>
  Hey there!
</div>
`;

exports[`Element with style props 3`] = `
<div
  style={
    Object {
      "color": "red",
    }
  }
>
  Hey there!
</div>
`;

exports[`StyleSheeet 1`] = `
<style
  dangerouslySetInnerHTML={
    Object {
      "__html": "
        body {
          color: red;
        }
      ",
    }
  }
/>
`;

exports[`Styler 1`] = `
<span
  style={
    Object {
      "color": "red",
    }
  }
>
  Red!
</span>
`;
